src/console/man/om_layer.1
.\" Text automatically generated by txt2man
.TH om_layer  "24 September 2013" "" ""
.SH NAME
\fBom_layer \fP- Layer utility functions for openModeller
\fB
.SH SYNOPSIS
.nf
.fam C
     \fBom_layer\fP [-] \fIv\fP \fB--version\fP | \fId\fP \fB--scan-dir\fP \fIDIRECTORY\fP [\fIs\fP \fB--result\fP \fIFILE\fP] | \fIl\fP \fB--check\fP \fILAYER_ID\fP [ \fB--log-level\fP \fILEVEL\fP ] [ \fIc\fP \fB--config-file\fP \fIFILE\fP ]

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
\fBom_layer\fP is a command line tool that can be used to either 1) check if a specified layer can be read using one of the openModeller drivers, or 2) recursively scan a specified directory to find openModeller-compatible layers.
.SH OPTIONS
.TP
.B
-\fIv\fP, \fB--version\fP
Display version info.
.TP
.B
-\fId\fP, \fB--scan-dir\fP
Scan this directory for layers.
.TP
.B
-\fIs\fP, \fB--result\fP
Store scan result in this file.
.TP
.B
-\fIl\fP, \fB--check\fP
Layer id to be checked.
.TP
.B
\fB--log-level\fP
openModeller log level: debug, warn, info or error. Defaults to "info".
.PP
-\fIc\fP, \fB--config-file\fP Configuration file for openModeller.
.SH AUTHORS
Renato De Giovanni <renato at cria dot org dot br>


ChangeLog
Release 1.4.0 (?-?-?)
-------------
SVN revision: ?

	* New CONSENSUS algorithm that can be used to generate models with different algorithms by converting each individual model into a binary model given a threshold and then "merging" them into a final combined model according to a specified level of agreement.
	* Upgraded gSOAP version being used by OMWS to 2.8.13.
	* Included library API versioning, starting with 0 from this release.
	* Moved some of the compiler definitions from CMake to internal #defines, so that client programs can include openModeller headers without worrying about adding specific definitions if they want to compile directly without using CMake.
	* New command-line tool om_evaluate to return raw model values given a set of points and an environmental scenario.
	* Fixed segfault when needNormalization was called on a deserialized SVM algorithm.
	* Modified om_pseudo command-line tool to also accept XML input (in this case also producing XML output), optionally storing it into a file. When a model is specified as a parameter, presences are generated inside the model, while absences are generated outside.

Release 1.3.0 (2012-07-23)
-------------
SVN revision: 5592

	* New version of OM-MAXENT compatible with Maxent 3.3.3e.
	* Changed the deserialization code to accept XML with namespace prefixes (they are now ignored if present).
	* Fixed bug in ANN algorithm which was limiting the number of input points to 10k (seg faulting otherwise).
	* New documentation about the web service interface (OMWS), aka SOAP interface.
	* New documentation about oM Server (server-side implementation of OMWS).
	* Fixed bug in oM Server getProgress operation which was sometimes returning 100% when the job was still unfinished.
	* New oM Server configuration/script allowing automatic synchronization of layers.
	* Several improvements in the OMWS Perl command-line client.
	* Several adjustments in the SWIG Python interface, exposing more methods and objects.
	* Fixed memory leak in the Environmental Distance algorithm when the Mahalanobis metric was used after model deserialization.
	* New WCS proxy raster driver that allows WCS rasters to be identified through strings in the format: wcs>wcs_service_url>coverage_id

Release 1.2.0 (2011-07-21)
--------------------------
SVN revision: 5330

	* Changed the ENFA algorithm to force marginality to be the first factor.  Changed broken stick distribution to work on specialisation factors only.  Add new parameter to enable log display of variable loadings for retained factors.
	* om_console now accepts a second parameter (after the request file) to specify the log level.
	* Changed the ENFA algorithm allowing absence data to be treated as background points, and log transforming the geomean calculation to avoid the occasional float overflow with large numbers of presences. Fixed a bug in the RetainVariation parameter (range changed from 50-100 to 0.5-1.0). Also fixed a deserialization error.
	* Compilation now requires CMake 2.6.4.
	* New algorithm Random Forests.
	* Web service interface now accepts a new optional parameter specifying the prediction threshold to be used when calculating area statistics during projection.
	* Renamed envelope element in the getProjectionMetadata response of the web service API to distinguish it from the subelement.
	* Added control to a maximum number of concurrent jobs on the modelling server.
	* Environmental values can now be loaded from the Sampler XML when provided.
	* Web service interface now accepts optional parameters: IgnoreAbsences for the confusion matrix and UseAbsencesAsBackground for the ROC curve calculations.
	* Added support to gzip compression in the Web service interface.
	* Changed the Environmental Distance algorithm to return chi-square distribution probabilities when the metric is Mahalanobis and the maximum distance is 1.
	* Changed the Maxent algorithm allowing absence data to be treated as background points. Also fixed memory deallocation issues.

Release 1.1.0 (2010-02-05)
--------------------------
SVN revision: 5114

	* Bugfix: Interquartile and envelope size adjustments were using wrong reference values for salinity and temperature in the Aquamaps algorithm.
	* Bugfix: All GARP algorithms were crashing when the abundance value associated with one of the occurrence points had more than two digits in the integer part.
	* Increased the default number of neurons in the hidden layer of the ANN algorithm.
	* New argument "s" in om_niche allowing model outputs to be scaled.
	* Re-implemented and re-enabled Maximum entropy algorithm based on the well-known Maxent.
	* Changed environmental distance algorithm forcing the Mahalanobis distance to use an identity matrix as the covariance matrix when there's only one input point.
	* Fixed small logical error in the CSM algorithm that could affect models with a large number of layers.
	* New algorithm (Niche Mosaic) implemented by Dr. Missae Yamamoto.
	* New algorithm (ENFA) kindly implemented by Dr. Chris Yesson.

Release 1.0 (2009-05-22)
------------------------
SVN revision: 5035

	* Changed Artificial Neural Networks algorithm to avoid exceeding a maximum number of iterations when training by minimum error (parameter Epoch is used to specify the maximum number of iterations).
	* Added new method in the ConfusionMatrix class to use the lowest probability among the training presences as the threshold.
	* Changed the modelling protocol to allow createModel and testModel requests to specify what statistics should be calculated.
	* Removed model statistics parameters from om_model since they can be specified as part of the XML request.
	* All command line programs now return the same openModeller version number.
	* Changed om_niche parameters (now using getopts) and created a man page for it.
	* Implemented custom log file in the modelling server.

trunk/openmodeller/src/console/om_layer.cpp
#include <openmodeller/om.hh>
#include <openmodeller/om_defs.hh>
#include <openmodeller/Log.hh>
#include <openmodeller/os_specific.hh>

#include "getopts/getopts.h"

#include "om_cmd_utils.hh"
#include "om_layer_utils.hh"

#include <fstream>   // file I/O for XML
#include <string>
#include <stdexcept>

using namespace std;

int main( int argc, char **argv ) {

  Options opts;
  int option;

  opts.addOption( "" , "log-level"  , "Set the log level (debug, warn, info, error)", true );
  opts.addOption( "v", "version"    , "Display version info"                        , false );
  opts.addOption( "d", "scan-dir"   , "Directory to be scanned for layers"          , true );
  opts.addOption( "s", "result"     , "File to store scan result"                   , true );
  opts.addOption( "l", "check"      , "Layer to be checked"                         , true );
  opts.addOption( "c", "config-file", "Configuration file for openModeller"         , true );

  std::string log_level("info");
  std::string scan_dir;
  std::string layer_id;
  std::string config_file;
  std::string result_file;

  if ( ! opts.parse( argc, argv ) ) {

    opts.showHelp( argv[0] ); 
    exit(0);
  }

  OpenModeller om;

  while ( ( option = opts.cycle() ) >= 0 ) {

    switch ( option ) {

      case 0:
        log_level = opts.getArgs( option );
        break;
      case 1:
        printf( "om_layer %s\n", om.getVersion().c_str() );
        printf("This is free software; see the source for copying conditions. There is NO\n");
        printf("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");
        exit(0);
        break;
      case 2:
        scan_dir = opts.getArgs( option );
        break;
      case 3:
        result_file = opts.getArgs( option );
        break;
      case 4:
        layer_id = opts.getArgs( option );
        break;
      case 5:
        config_file = opts.getArgs( option );
        break;
      default:
        break;
    }
  }

  // Check parameters

  if ( (scan_dir.empty() && layer_id.empty()) || (!scan_dir.empty() && !layer_id.empty()) ) {

    printf( "Please specify one and only one of the parameters: --check layer_id, --scan-dir directory\n");
    exit(1);
  }

  // Log stuff
  Log::Level level_code = getLogLevel( log_level );

  Log::instance()->setLevel( level_code );

  // Config file
  if ( ! config_file.empty() ) {

    Settings::loadConfig( config_file );
  }

  if ( layer_id.size() > 0 ) {

    try {

      Raster * r = RasterFactory::instance().create( layer_id, 0 );

      delete r;
    }
    catch ( std::runtime_error e ) {

      printf( "Failed to process raster %s:\n\n%s\n", layer_id.c_str(), e.what() );
      exit(1);
    }
  }
  else {

    // Here we assume scan-dir was specified

    ostringstream oss;

    int seq = 1;

    if ( ! readDirectory( scan_dir.c_str(), "Layers", oss, 0, &seq ) ) {

      printf( "Failed to scan specified directory\n" );
      exit(1);
    }

    std::cerr << flush;

    // Write result to file, if requested
    if ( ! result_file.empty() ) {

      ofstream file( result_file.c_str() );
      file << oss.str();
      file.close();
    }
    else {

      // Otherwise send it to stdout
      std::cout << oss.str().c_str() << endl << flush;
    }
  }

  return 0;
}
